### PR TITLE
[Clang] [Sema] Do not attempt to dump the layout of dependent types when `-fdump-record-layouts-complete` is passed

### LINKED ISSUE #83671: [Clang] 19.0.0 SIGSEGV due to NULL Pointer dereference caused by receiving an NULL `BaseDecl` at `EmptySubobjectMap::ComputeEmptySubobjectSizes()`
Note: This is related to dump record layout feature with option `-Xclang -fdump-record-layouts-complete`.

When I want to use that option to dump `class` and `struct`'s layout, the clang crashes with SIGSEGV signal.

Crashsite Screenshot:
crashsite

Through my simple debugging, I found the problem might related to `EmptySubobjectMap::ComputeEmptySubobjectSizes()` in `clang/lib/AST/RecordLayoutBuilder.cpp`.

```cpp
void EmptySubobjectMap::ComputeEmptySubobjectSizes() {
  // Check the bases.
  for (const CXXBaseSpecifier &Base : Class->bases()) {
    const CXXRecordDecl *BaseDecl = Base.getType()->getAsCXXRecordDecl(); // [!]

    CharUnits EmptySize;
    const ASTRecordLayout &Layout = Context.getASTRecordLayout(BaseDecl); // [1]
    if (BaseDecl->isEmpty()) {
      // If the class decl is empty, get its size.
      EmptySize = Layout.getSize();
    } else {
      // Otherwise, we get the largest empty subobject for the decl.
      EmptySize = Layout.getSizeOfLargestEmptySubobject();
    }

    if (EmptySize > SizeOfLargestEmptySubobject)
      SizeOfLargestEmptySubobject = EmptySize;
  }

  // Check the fields.
  for (const FieldDecl *FD : Class->fields()) {
    const RecordType *RT =
        Context.getBaseElementType(FD->getType())->getAs<RecordType>();

    // We only care about record types.
    if (!RT)
      continue;

    CharUnits EmptySize;
    const CXXRecordDecl *MemberDecl = RT->getAsCXXRecordDecl();
    const ASTRecordLayout &Layout = Context.getASTRecordLayout(MemberDecl);
    if (MemberDecl->isEmpty()) {
      // If the class decl is empty, get its size.
      EmptySize = Layout.getSize();
    } else {
      // Otherwise, we get the largest empty subobject for the decl.
      EmptySize = Layout.getSizeOfLargestEmptySubobject();
    }

    if (EmptySize > SizeOfLargestEmptySubobject)
      SizeOfLargestEmptySubobject = EmptySize;
  }
}
```

At position `[!]`, `Base.getType()->getAsCXXRecordDecl()` will return a NULL pointer then causes crash in `Context.getASTRecordLayout(BaseDecl)` at position `[1]`.

Nullptr Introduce Site Screenshot:
nullptr-introduce

This is crash site, including the crash backtrace, preprocessed source, and associated run script.
[crashsite.zip](https://github.com/llvm/llvm-project/files/14469084/crashsite.zip)

Thanks!

### LINKED ISSUE #83684: [clang] 19.0.0 aborts due to Assertion `!T->isDependentType() && "should not see dependent types here"` failed in `tools/clang/include/clang/AST/TypeNodes.inc`
Compile the following `1.cpp` with command line `clang++ -Xclang -fdump-record-layouts-complete 1.cpp` cause a clang abort:
```cpp
template <class Pointer>
struct AllocationResult {
  Pointer ptr = nullptr;
  int count = 0;
};
int main(){}
```

The attachment is crash backtrace, preprocessed source, and associated run script.
[crashsite.zip](https://github.com/llvm/llvm-project/files/14470166/crashsite.zip)

### PR DESCRIPTION
**Pull request by Sirraide**
We were crashing on trying to print the layout of an uninstantiated template, which obviously doesn’t make sense.

This fixes #83684.

### CODE REVIEW

File: clang/lib/AST/Decl.cpp [hunk 1]
```diff
@@ -5042,7 +5042,7 @@ void RecordDecl::completeDefinition() {
 
   // Layouts are dumped when computed, so if we are dumping for all complete
   // types, we need to force usage to get types that wouldn't be used elsewhere.
-  if (Ctx.getLangOpts().DumpRecordLayoutsComplete)
+  if (Ctx.getLangOpts().DumpRecordLayoutsComplete && !isDependentType())
```
Comment (shafik) [inline clang/lib/AST/Decl.cpp:5045]: Can you update the comment above to reflect this change.

**Review by shafik — COMMENTED** [review]
I think this makes sense but I would like another set of eyes.

File: clang/lib/AST/Decl.cpp (hunk 1 above)
Comment (Sirraide) [reply clang/lib/AST/Decl.cpp:5045]: Sure; I was going to do that but I felt it wasn’t really ‘necessary’, but it can’t hurt.

**Comment by Endilll** [comment]
This might also fix #83671

**Comment by Sirraide** [comment]
> This might also fix #83671

From what I saw, that’s a different problem, and there’s a different pr for that already, but I’ll double-check.

**Comment by Sirraide** [comment]
Oh, apparently, there’s *yet another* problem here; it seems we should also check for `isInvalidDecl()`.

**Comment by bjrjk** [comment]
> > This might also fix #83671
> 
> From what I saw, that’s a different problem, and there’s a different pr for that already, but I’ll double-check.

Seems not same problem, root cause are different.

**Comment by Sirraide** [comment]
> This might also fix #83671

Actually, it seems to fix your reduced test case, hmm...

**Comment by Sirraide** [comment]
So yeah, this pr indeed also seems to fix #83671 as well.

**Review by AaronBallman — APPROVED** [review]
LGTM!

**Comment by Sirraide** [comment]
Apparently, buildbot is mad at me (<https://lab.llvm.org/buildbot/#/builders/165/builds/49741>)?

I’m not entirely sure how a fix that affects only `-fdump-record-layouts-complete` could have caused this. Is this a spurious failure or due to something unrelated, or is it worth investigating this?

**Pull request merged**